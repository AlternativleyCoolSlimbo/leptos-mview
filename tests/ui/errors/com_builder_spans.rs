//! Testing that there are no errors that cause the entire macro to error (i.e.
//! call-site error)

use leptos::*;
use leptos_mview::mview;

fn missing_args() {
    // missing `key` attribute
    _ = mview! {
        For each=[[1, 2, 3]] |i| { {i} }
    };
}

fn incorrect_arg_value() {
    // Show takes `bool` not `&str`
    _ = mview! {
        Show when={"no"} {
            "hi"
        }
    };
}

fn missing_closure() {
    _ = mview! {
        Show when={true} {
            "hi"
        }
    };
}

fn incorrect_closure() {
    #[component]
    fn Thing(label: &'static str) -> impl IntoView { label }

    // `label` is not a closure
    _ = mview! {
        Thing label=[false];
    };
}

fn incorrect_closure_to_children() {
    #[component]
    fn Thing(children: Children) -> impl IntoView { children() }

    let s = String::new();
    // `children` does not take a closure
    _ = mview! {
        Thing |s| { "hello" }
    };
}

fn missing_closure_to_children() {
    // thought it would make an error at the `.children`, but it seem to accept it
    // and errors at the tag name instead. this test is just for notifying in case
    // this ever changes.
    _ = mview! {
        Await future=[async { 3 }] { "no args" }
    };
}

fn main() {}
