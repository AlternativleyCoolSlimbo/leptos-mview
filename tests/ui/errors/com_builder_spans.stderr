warning: use of deprecated method `leptos::ForPropsBuilder::<IF, I, T, EF, N, KF, K, ((IF,), (), __children)>::build`: Missing required field key
  --> tests/ui/errors/com_builder_spans.rs:10:9
   |
10 |         For each=[[1, 2, 3]] |i| { {i} }
   |         ^^^
   |
   = note: `#[warn(deprecated)]` on by default

error[E0061]: this method takes 1 argument but 0 arguments were supplied
  --> tests/ui/errors/com_builder_spans.rs:10:9
   |
10 |         For each=[[1, 2, 3]] |i| { {i} }
   |         ^^^ an argument of type `ForPropsBuilder_Error_Missing_required_field_key` is missing
   |
note: method defined here
  --> $CARGO/leptos-0.5.2/src/for_loop.rs
   |
   | #[component(transparent)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `::leptos::typed_builder_macro::TypedBuilder` (in Nightly builds, run with -Z macro-backtrace for more info)
help: provide the argument
   |
10 |         For(/* ForPropsBuilder_Error_Missing_required_field_key */) each=[[1, 2, 3]] |i| { {i} }
   |            ++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: expected a `Fn()` closure, found `str`
  --> tests/ui/errors/com_builder_spans.rs:17:14
   |
17 |         Show when={"no"} {
   |              ^^^^ expected an `Fn()` closure, found `str`
   |
   = help: the trait `std::ops::Fn<()>` is not implemented for `str`
   = note: wrap the `str` in a closure with no arguments: `|| { /* code */ }`
   = note: required for `&str` to implement `FnOnce<()>`
note: required by a bound in `ShowPropsBuilder::<W, (__children, (), __fallback)>::when`
  --> $CARGO/leptos-0.5.2/src/show.rs
   |
   |     when: W,
   |     ---- required by a bound in this associated function
...
   |     W: Fn() -> bool + 'static,
   |                ^^^^ required by this bound in `ShowPropsBuilder::<W, (__children, (), __fallback)>::when`

error[E0277]: expected a `Fn()` closure, found `str`
  --> tests/ui/errors/com_builder_spans.rs:17:9
   |
17 |         Show when={"no"} {
   |         ^^^^ expected an `Fn()` closure, found `str`
   |
   = help: the trait `std::ops::Fn<()>` is not implemented for `str`
   = note: wrap the `str` in a closure with no arguments: `|| { /* code */ }`
   = note: required for `&str` to implement `std::ops::Fn<()>`
note: required by a bound in `leptos::Show`
  --> $CARGO/leptos-0.5.2/src/show.rs
   |
   | pub fn Show<W>(
   |        ---- required by a bound in this function
...
   |     W: Fn() -> bool + 'static,
   |        ^^^^^^^^^^^^ required by this bound in `Show`

error[E0599]: the method `children` exists for struct `ShowPropsBuilder<&str, ((), (&str,), ())>`, but its trait bounds were not satisfied
  --> tests/ui/errors/com_builder_spans.rs:18:13
   |
16 |       _ = mview! {
   |  _________-
17 | |         Show when={"no"} {
18 | |             "hi"
   | |            -^^^^ method cannot be called on `ShowPropsBuilder<&str, ((), (&str,), ())>` due to unsatisfied trait bounds
   | |____________|
   |
   |
   = note: the following trait bounds were not satisfied:
           `<&str as FnOnce<()>>::Output = bool`
           `&str: std::ops::Fn<()>`
           `&str: FnMut<()>`
           which is required by `&str: std::ops::Fn<()>`
           `&str: FnOnce<()>`
           which is required by `&str: std::ops::Fn<()>`

error[E0308]: mismatched types
  --> tests/ui/errors/com_builder_spans.rs:29:21
   |
29 |         Thing label=[false];
   |               ----- ^^^^^^^ expected `&str`, found closure
   |               |
   |               arguments to this method are incorrect
   |
   = note: expected reference `&'static str`
                found closure `{closure@$DIR/tests/ui/errors/com_builder_spans.rs:29:21: 29:28}`
note: method defined here
  --> tests/ui/errors/com_builder_spans.rs:25:14
   |
25 |     fn Thing(label: &'static str) -> impl IntoView { label }
   |              ^^^^^--------------

error[E0308]: mismatched types
  --> tests/ui/errors/com_builder_spans.rs:40:15
   |
40 |         Thing |s| { "hello" }
   |               ^^^ expected `Box<dyn FnOnce() -> Fragment>`, found closure
   |
   = note: expected struct `Box<(dyn FnOnce() -> Fragment + 'static)>`
             found closure `{closure@$DIR/tests/ui/errors/com_builder_spans.rs:40:15: 40:18}`

error[E0283]: type annotations needed
  --> tests/ui/errors/com_builder_spans.rs:49:9
   |
49 |         Await future=[async { 3 }] { "no args" }
   |         ^^^^^ cannot infer type of the type parameter `V` declared on the function `Await`
   |
   = note: cannot satisfy `_: leptos::IntoView`
   = help: the following types implement trait `leptos::IntoView`:
             bool
             char
             isize
             i8
             i16
             i32
             i64
             i128
           and $N others
note: required by a bound in `leptos::Await`
  --> $CARGO/leptos-0.5.2/src/await_.rs
   |
   | pub fn Await<T, Fut, FF, VF, V>(
   |        ----- required by a bound in this function
...
   |     V: IntoView,
   |        ^^^^^^^^ required by this bound in `Await`
help: consider specifying the generic arguments
   |
49 |         Await::<i32, _, _, VF, V> future=[async { 3 }] { "no args" }
   |              ++++++++++++++++++++

error[E0283]: type annotations needed
  --> tests/ui/errors/com_builder_spans.rs:49:9
   |
49 |         Await future=[async { 3 }] { "no args" }
   |         ^^^^^ cannot infer type of the type parameter `VF` declared on the function `Await`
   |
   = note: multiple `impl`s satisfying `_: ToChildren<{closure@$DIR/tests/ui/errors/com_builder_spans.rs:49:38: 49:47}>` found in the `leptos` crate:
           - impl<F> ToChildren<F> for Box<(dyn FnMut() -> Fragment + 'static)>
             where <F as FnOnce<()>>::Output == Fragment, F: FnMut<()>, F: 'static;
           - impl<F> ToChildren<F> for Box<(dyn FnOnce() -> Fragment + 'static)>
             where <F as FnOnce<()>>::Output == Fragment, F: FnOnce<()>, F: 'static;
           - impl<F> ToChildren<F> for Box<(dyn std::ops::Fn() -> Fragment + 'static)>
             where <F as FnOnce<()>>::Output == Fragment, F: std::ops::Fn<()>, F: 'static;
           - impl<F> ToChildren<F> for Rc<(dyn std::ops::Fn() -> Fragment + 'static)>
             where <F as FnOnce<()>>::Output == Fragment, F: std::ops::Fn<()>, F: 'static;
help: consider specifying the generic arguments
   |
49 |         Await::<i32, _, _, VF, V> future=[async { 3 }] { "no args" }
   |              ++++++++++++++++++++
